### PR TITLE
Add ChronosTime

### PR DESCRIPTION
**Pull request by othercorey**
refs https://github.com/cakephp/chronos/issues/102

Adds time representation that attempts to mimic the time parsing/behavior of PHP's DateTime.

### CODE REVIEW

**Comment by othercorey** [comment]
@markstory Dos this seem a reasonable direction to go?

File: src/ChronosTime.php [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+
+    /**
+     * @var int
+     */
+    protected int $microseconds;
+
+    /**
+     * @param \Cake\Chronos\Chronos|\Cake\Chronos\ChronosDate|\Cake\Chronos\ChronosTime|\DateTimeInterface|string|int $time Time string or microseconds
+     */
+    public function __construct(Chronos|ChronosDate|ChronosTime|DateTimeInterface|string|int $time = 0)
```
Comment (markstory) [inline src/ChronosTime.php]:
```suggestion
    public function __construct(Chronos|ChronosTime|DateTimeInterface|string|int $time = 0)
```

I don't know how you could convert a Date into Time.

File: src/ChronosTime.php [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+        return $microseconds;
+    }
+
+    /**
+     * @return int
+     */
+    public function getMicroseconds(): int
+    {
+        return $this->microseconds % self::MICRO_PER_SECOND;
```
Comment (markstory) [inline src/ChronosTime.php]: Should this just return the microseconds?

**Review by markstory — COMMENTED** [review]
This is looking like a good start to me.

File: src/ChronosTime.php (hunk 2 above)
Comment (othercorey) [reply src/ChronosTime.php]: This is DateTime microseconds. The same value you get with `u` format specifier.

Comment (markstory) [reply src/ChronosTime.php]: :man_facepalming: my bad.

File: src/ChronosTime.php [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+            throw new InvalidArgumentException('Unable to parse time string.');
+        }
+
+        $microseconds = $matches[1] * self::MICRO_PER_HOUR;
+        $microseconds += $matches[2] * self::MICRO_PER_MINUTE;
+        if (isset($matches[3])) {
+            $microseconds += $matches[3] * self::MICRO_PER_SECOND;
+        }
+        if (isset($matches[4])) {
+            $microseconds += substr($matches[4], 0, 6);
```
Comment (markstory) [inline src/ChronosTime.php]:
```suggestion
            $microseconds += (int)substr($matches[4], 0, 6);
```

File: src/ChronosTime.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+    protected static function parseTime(string $time): int
+    {
+        if (!preg_match('/^\s*(\d{1,2})[:.](\d{1,2})(?:[:.](\d{1,2}))(?:[.](\d+))\s*$/', $time, $matches)) {
+            throw new InvalidArgumentException('Unable to parse time string.');
+        }
+
+        $microseconds = $matches[1] * self::MICRO_PER_HOUR;
+        $microseconds += $matches[2] * self::MICRO_PER_MINUTE;
+        if (isset($matches[3])) {
+            $microseconds += $matches[3] * self::MICRO_PER_SECOND;
```
Comment (markstory) [inline src/ChronosTime.php]:
```suggestion
            $microseconds += (int)$matches[3] * self::MICRO_PER_SECOND;
```

File: src/ChronosTime.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+     * @return int
+     */
+    protected static function parseTime(string $time): int
+    {
+        if (!preg_match('/^\s*(\d{1,2})[:.](\d{1,2})(?:[:.](\d{1,2}))(?:[.](\d+))\s*$/', $time, $matches)) {
+            throw new InvalidArgumentException('Unable to parse time string.');
+        }
+
+        $microseconds = $matches[1] * self::MICRO_PER_HOUR;
+        $microseconds += $matches[2] * self::MICRO_PER_MINUTE;
```
Comment (markstory) [inline src/ChronosTime.php]:
```suggestion
        $microseconds += (int)$matches[2] * self::MICRO_PER_MINUTE;
```

File: src/ChronosTime.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+     * @param string $time Time string (HH:mm(:ss(.u)))
+     * @return int
+     */
+    protected static function parseTime(string $time): int
+    {
+        if (!preg_match('/^\s*(\d{1,2})[:.](\d{1,2})(?:[:.](\d{1,2}))(?:[.](\d+))\s*$/', $time, $matches)) {
+            throw new InvalidArgumentException('Unable to parse time string.');
+        }
+
+        $microseconds = $matches[1] * self::MICRO_PER_HOUR;
```
Comment (markstory) [inline src/ChronosTime.php]:
```suggestion
        $microseconds = (int)$matches[1] * self::MICRO_PER_HOUR;
```

File: src/ChronosTime.php [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+    }
+
+    /**
+     * @return int
+     */
+    public function getSeconds(): int
+    {
+        $secondsInMicro = $this->microseconds % self::MICRO_PER_MINUTE - $this->microseconds % self::MICRO_PER_SECOND;
+
+        return $secondsInMicro / self::MICRO_PER_SECOND;
```
Comment (garas) [inline src/ChronosTime.php]:
```suggestion
        return intdiv($secondsInMicro, self::MICRO_PER_SECOND);
```
\+ other divisions

Comment (othercorey) [reply src/ChronosTime.php]: I was going to use the TypeError to catch any conversion issues, but this is a good idea. However, I have to modify the implementation to support other features.

File: src/ChronosTime.php [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+    }
+
+    /**
+     * @param string $time Time string in the format HH[:.]mm or HH[:.]mm[:.]ss.u
+     * @return int
+     */
+    protected static function parseTime(string $time): int
+    {
+        if (!preg_match('/^\s*(\d{1,2})[:.](\d{1,2})(?|[:.](\d{1,2})[.](\d+)|[:.](\d{1,2}))?\s*$/', $time, $matches)) {
+            throw new InvalidArgumentException('Unable to parse time string.');
```
Comment (ADmad) [inline src/ChronosTime.php]:
```suggestion
            throw new InvalidArgumentException('Unable to parse time string. It should be in the format HH[:.]mm or HH[:.]mm[:.]ss.u');
```

File: src/ChronosTime.php [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+        }
+
+        $hours = (int)$matches[1];
+        $minutes = (int)$matches[2];
+        $seconds = (int)($matches[3] ?? 0);
+        $microseconds = (int)substr($matches[4] ?? '', 0, 6);
+
+        if ($hours > 23 || $minutes > 59 || $seconds > 59 || $microseconds > 999_999) {
+            throw new InvalidArgumentException('Unable to parse time string.');
```
Comment (ADmad) [inline src/ChronosTime.php]:
```suggestion
            throw new InvalidArgumentException('Unable to parse time string. Invalid value provided.');
```

**Pull request merged**